tests/network/services.go

```diff
@@ -50,182 +49,81 @@ import (
-				console.LoginToCirros)
-		}
-
 		BeforeEach(func() {
 			libnet.SkipWhenClusterNotSupportIpv4()
-			subdomain := "vmi"
-			hostname := "inbound"
 
-			inboundVMI = createReadyVMIWithBridgeBindingAndExposedService(hostname, subdomain)
-			tests.StartTCPServer(inboundVMI, servicePort, console.LoginToCirros)
-		})
+			inboundVMI = libvmi.NewCirros(
+				libvmi.WithInterface(libvmi.InterfaceDeviceWithBridgeBinding(v1.DefaultPodNetwork().Name)),
+				libvmi.WithNetwork(v1.DefaultPodNetwork()),
+				libvmi.WithLabel(selectorLabelKey, selectorLabelValue),
+				libvmi.WithSubdomain("vmi"),
+				libvmi.WithHostname("inbound"),
+			)
+			var err error
+			inboundVMI, err = kubevirt.Client().VirtualMachineInstance(util.NamespaceTestDefault).Create(context.Background(), inboundVMI)
+			Expect(err).ToNot(HaveOccurred())
 
-		AfterEach(func() {
-			Expect(inboundVMI).NotTo(BeNil(), "the VMI object must exist in order to be deleted.")
-			cleanupVMI(virtClient, inboundVMI)
+			inboundVMI = libwait.WaitUntilVMIReady(inboundVMI, console.LoginToCirros)
+			tests.StartTCPServer(inboundVMI, servicePort, console.LoginToCirros)
 		})
 
 		Context("with a service matching the vmi exposed", func() {
-			var jobCleanup func() error
+			const serviceName = "myservice"
+			virtClient := kubevirt.Client()
 
 			BeforeEach(func() {
-				serviceName = "myservice"
-
 				service := netservice.BuildSpec(serviceName, servicePort, servicePort, selectorLabelKey, selectorLabelValue)
-				_, err := virtClient.CoreV1().Services(inboundVMI.Namespace).Create(context.Background(), service, k8smetav1.CreateOptions{})
+				var err error
+				service, err = virtClient.CoreV1().Services(inboundVMI.Namespace).Create(context.Background(), service, k8smetav1.CreateOptions{})
 				Expect(err).ToNot(HaveOccurred())
-			})
-
-			AfterEach(func() {
-				Expect(cleanupService(inboundVMI.GetNamespace(), serviceName)).To(Succeed(), cleaningK8sv1ServiceShouldSucceed)
-			})
-
-			AfterEach(func() {
-				Expect(jobCleanup).NotTo(BeNil(), cleaningK8sv1JobFuncShouldExist)
-				Expect(jobCleanup()).To(Succeed(), cleaningK8sv1JobShouldSucceed)
-				jobCleanup = nil
+				deferServiceCleanup(service.Namespace, service.Name)
 			})
 
 			It("[test_id:1547] should be able to reach the vmi based on labels specified on the vmi", func() {
-				var err error
-
-				jobCleanup, err = assertConnectivityToService(serviceName, inboundVMI.Namespace, servicePort)
+				job, err := runConnectivityJobToService(serviceName, inboundVMI.Namespace, servicePort)
 				Expect(err).NotTo(HaveOccurred(), expectConnectivityToExposedService)
+				deferJobCleanup(job.Namespace, job.Name)
 			})
 
 			It("[test_id:1548] should fail to reach the vmi if an invalid servicename is used", func() {
-				var err error
-
-				jobCleanup, err = assertNoConnectivityToService("wrongservice", inboundVMI.Namespace, servicePort)
+				job, err := runNoConnectivityJobToService("wrongservice", inboundVMI.Namespace, servicePort)
```

Comment:
Going back to deferred cleanup: I think that this test code would be more tight if you:
1. create job
2. defer its cleanup
3. expect that it fails

You would not allow a 1-3-2 order in production code; you shouldn't have it in tests either.

Reply:
Now that this was explained offline, I get it.

The problem here is that `runNoConnectivity..` is doing two things: Created the job and waits for it to finish. We should defer the cleanup if it was created, regardless if the "wait" succeeded or failed.
In the scenario that the "wait" failed, the explicit cleanup will not be registered, which is wrong.

Fixed.